cli/src/org/partiql/shell/ShellHighlighter.kt

```diff
@@ -11,32 +11,134 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
  * language governing permissions and limitations under the License.
  */
-
 package org.partiql.shell
 
-import org.jline.builtins.Nano.SyntaxHighlighter
+import com.amazon.ion.IonString
+import com.amazon.ion.system.IonSystemBuilder
 import org.jline.reader.Highlighter
 import org.jline.reader.LineReader
 import org.jline.utils.AttributedString
+import org.jline.utils.AttributedStringBuilder
 import org.jline.utils.AttributedStyle
+import org.partiql.lang.errors.Property
+import org.partiql.lang.syntax.ParserException
+import org.partiql.lang.syntax.SqlLexer
+import org.partiql.lang.syntax.SqlParser
+import org.partiql.lang.syntax.Token
+import org.partiql.lang.syntax.TokenType
 import java.io.PrintStream
-import java.nio.file.Path
 import java.util.regex.Pattern
 
 private val SUCCESS: AttributedStyle = AttributedStyle.DEFAULT.foreground(AttributedStyle.GREEN)
 private val ERROR: AttributedStyle = AttributedStyle.DEFAULT.foreground(AttributedStyle.RED)
 private val INFO: AttributedStyle = AttributedStyle.DEFAULT
 private val WARN: AttributedStyle = AttributedStyle.DEFAULT.foreground(AttributedStyle.YELLOW)
 
-class ShellHighlighter(private val syntaxFile: Path) : Highlighter {
+internal class ShellHighlighter() : Highlighter {
+
+    private val ion = IonSystemBuilder.standard().build()
+    private val lexer = SqlLexer(ion)
+    private val parser = SqlParser(ion)
+
+    /**
+     * Returns a highlighted string by passing the [input] string through the [lexer] and [parser] to identify and
+     * highlight tokens
+     */
+    override fun highlight(reader: LineReader, input: String): AttributedString {
+        if (input.isEmpty()) return AttributedString(input)
+
+        var builder = AttributedStringBuilder()
+        builder.append(AttributedString(input))
+
+        // Map Between Line Number and Index
+        val lineIndexesMap = mutableMapOf<Int, Int>()
+        var currentLine = 0
+        lineIndexesMap[currentLine] = -1
+        input.forEachIndexed { index, char ->
+            if (char == '\n') lineIndexesMap[++currentLine] = index
+        }
 
-    private val syntax = SyntaxHighlighter.build(syntaxFile, "PartiQL")
+        // Get Tokens
+        val tokens: List<Token>
+        try {
+            tokens = lexer.tokenize(input)
+        } catch (e: Exception) {
+            return AttributedString(input, AttributedStyle().foreground(AttributedStyle.RED))
+        }
 
-    override fun highlight(reader: LineReader, line: String): AttributedString = syntax.highlight(line)
+        // Replace Token Colors
+        tokens.forEach { builder = createAttributeStringBuilder(it, lineIndexesMap, builder, false) }
```

Comment:
This confuses me. Why are we continually updating the value of builder almost like a fold?

Why not do
```
val builder = AttributedStringBuilder()
tokens.forEach { addTokenToBuilder(input, lineIndexesMap, builder, false)
```

Reply:
I guess I'm confused why the builder is returned

Reply:
Modifying to build a string. The only caveat is for parsing errors which needs to highlight a specific word red. You'll like the next revision posting soon. Thanks for the comment. Makes sense.